stable/database/files/nuosm

```diff
@@ -33,312 +39,499 @@ if [ $crashcount -ge 1 ]; then
 
   # if the latest backup was not by my group, then wait to allow an SM from the latest group to start first
   if [ "$latest_group" != "$NUODB_BACKUP_GROUP" ]; then
-    echo ":latest backup was not made by my group $NUODB_BACKUP_GROUP - waiting to allow an SM from $latest_group to start..."
+    log ":latest backup was not made by my group $NUODB_BACKUP_GROUP - waiting to allow an SM from $latest_group to start..."
 
     # try for 30 seconds, sleeping every 2 seconds
     for retry in {1..30..2}; do
-      sm="$($NUOCMD get value --key $first_req )"
+      sm="$(nuocmd get value --key $first_req )"
       [ -n "$sm" ] && break
       sleep 2
     done
 
-    [ -n "$sm" ] && echo "Primary restore SM='${sm}'..." || echo "No Primary restore SM found - attempting restore from $NUODB_BACKUP_GROUP..."
+    [ -n "$sm" ] && log "Primary restore SM='${sm}'..." || log "No Primary restore SM found - attempting restore from $NUODB_BACKUP_GROUP..."
   fi
 fi
 
 # resolve the latest backup for the specified backup group
 if [ "$restore_source" = ":latest" -o "$restore_source" = ":group-latest" ]; then
-  [ -n "$NUODB_DEBUG" ] && echo "Resolving restore '$restore_source'..."
+
+  trace "retrieving :group-latest from nuorestore"
+
+  log "Resolving restore '$restore_source'..."
   restore_source=$( nuobackup --type report-latest --db-name $DB_NAME --group $NUODB_BACKUP_GROUP )
-  [ -n "$NUODB_DEBUG" ] && echo "Latest restore for $NUODB_BACKUP_GROUP resolved to $restore_source"
+  log "Latest restore for $NUODB_BACKUP_GROUP resolved to $restore_source"
 fi
 
-if [ -n "$restore_requested" -a -n "$restore_source" ]; then
+atomCount=$( ls -l $DB_DIR/*.atm 2>/dev/null | wc -l )
+catalogCount=$( ls -l $DB_DIR/*.cat 2>/dev/null | wc -l )
+log "restore_source=$restore_source; restore_requested=$restore_requested; path=$DB_DIR; atoms=${atomCount}; catalogs=${catalogCount}"
+
+wotIdid=""
 
+if [ -n "$restore_requested" -a -n "$restore_source" ]; then
   # work out who is the first one in
-  $NUOCMD set value --key $first_req --value $HOSTNAME --expected-value ''
-  first_in="$( $NUOCMD get value --key $first_req )"
+  trace "trying to reserve first-in"
+  nuocmd set value --key $first_req --value $HOSTNAME --expected-value ''
+  first_in="$( nuocmd get value --key $first_req )"
 
-  echo "First-in = $first_in"
+  wotIdid=""
+
+  log "First-in = $first_in"
 
   # if I got in first - perform the restore
   if [ "$first_in" = "$HOSTNAME" ]; then
 
-    echo "I am first-in: $first_in == $HOSTNAME"
-
-    # [ -z "$restore_source" ] && die -1 "There is no valid LATEST backup - restore of LATEST failed."
+    log "I am first-in: $first_in == $HOSTNAME"
 
     # take ownership of the SM startup semaphore
-    $NUOCMD set value --key $startup_key/$DB_NAME --value $HOSTNAME --unconditional
+    trace "Take ownership of SM startup semaphore"
+    nuocmd set value --key $startup_key/$DB_NAME --value $HOSTNAME --unconditional
```

Comment:
What is this KV wrangling for? This is probably a question for @NikTJ777.

Reply:
As far as I can tell, there are two main KV keys used for synchronization:
- `$NUODB_RESTORE_REQUEST_PREFIX/$DB_NAME/first` - this is held by the SM which will perform restore. Any other SM will wait for the "first" SM to go into `RUNNING` state before it starts. 
- `$startup_key/$DB_NAME` - this is held by every SM when they start so that SMs start in sequential order so that we don't have more than one SYNCing SM at a time. This is probably to address an old issue where we had problems when more than 2 SMs start SYNCing.